### PR TITLE
[JENKINS-61438] Demonstration of IAE from Stapler

### PR DESCRIPTION
**Pull request by nfalco79**
The new test case fails with
```
java.lang.IllegalArgumentException: Unable to convert to class java.lang.Object
	at org.kohsuke.stapler.RequestImpl$TypePair.convertJSON(RequestImpl.java:738)
	at org.kohsuke.stapler.RequestImpl.bindJSON(RequestImpl.java:478)
	at org.kohsuke.stapler.RequestImpl.injectSetters(RequestImpl.java:834)
	at org.kohsuke.stapler.RequestImpl.instantiate(RequestImpl.java:784)
	at org.kohsuke.stapler.RequestImpl.access$200(RequestImpl.java:83)
	at org.kohsuke.stapler.RequestImpl$TypePair.convertJSON(RequestImpl.java:678)
Caused: java.lang.IllegalArgumentException: Failed to instantiate class hudson.model.ParameterDefinition from {"name":"CHOOSE","choices":"a\nb","description":"choose one","stapler-class":"hudson.model.ChoiceParameterDefinition","$class":"hudson.model.ChoiceParameterDefinition"}
	at org.kohsuke.stapler.RequestImpl$TypePair.convertJSON(RequestImpl.java:681)
```

@jglick Now I will try to use the incremental build of stapler to prove the fix is valid

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -78,6 +78,10 @@
       <id>repo.jenkins-ci.org</id>
       <url>https://repo.jenkins-ci.org/public/</url>
     </repository>
+    <repository>
+      <id>maven.jenkins-ci.org</id>
+      <url>https://repo.jenkins-ci.org/incrementals/</url>
+    </repository>
```
Comment (jglick) [inline pom.xml]:
```suggestion
```

see https://github.com/jenkinsci/branch-api-plugin/blob/2b1b650338e8ff2daf0f997ee3b231c5cbe25aa8/.mvn/maven.config#L1

Comment (nfalco79) [reply pom.xml]: ok but I would not merge this PR. This PR is just to demostrate what you ask in the stapler repository. This was only for demo purpose.

Comment (jglick) [reply pom.xml]: Sure, but assuming the base PR is merged we can just amend this to `@Ignore` the test with a `TODO` comment reminding the maintainer to reënable it once the fix makes it into a suitable `jenkins.version`.

Comment (nfalco79) [reply pom.xml]: You mean I totally remove the changes to pom.xml and add the
`@Ignore("Enable this test when project update jenkins version >= XYZ")` ?

Comment (jglick) [reply pom.xml]: Yes, something like that, so that the PR becomes safe to merge at any time.

Comment (nfalco79) [reply pom.xml]: done

Comment (nfalco79) [reply pom.xml]: also restore import packages order

**Comment by jglick** [comment]
For reference: https://github.com/stapler/stapler/issues/181 https://github.com/stapler/stapler/pull/182

File: src/test/java/jenkins/branch/ParameterDefinitionBranchPropertyTest.java [hunk 2]
```diff
@@ -116,6 +120,40 @@ public void jobDecorator_Job_implementing_ParameterizedJob() throws Exception {
...
+            BranchSource source = new BranchSource(new MockSCMSource(c, "foo", new MockSCMDiscoverBranches()));
+            ParameterDefinitionBranchPropertyImpl instance = new ParameterDefinitionBranchPropertyImpl();
+            instance.setParameterDefinitions(Collections.<ParameterDefinition>singletonList(
+                    new ChoiceParameterDefinition("CHOOSE", new String[] { "a", "b" }, "choose one")
+                    ));
+            source.setStrategy(new DefaultBranchPropertyStrategy(new BranchProperty[]{
+                                                                                      instance, new BasicDummyStepBranchProperty()
+            }));
+            prj.getSourcesList().add(source);
+            r.configRoundtrip(prj);
```
Comment (jglick) [inline src/test/java/jenkins/branch/ParameterDefinitionBranchPropertyTest.java:141]: Is this where the test fails?

Comment (nfalco79) [reply src/test/java/jenkins/branch/ParameterDefinitionBranchPropertyTest.java:141]: I do not remember but I suppose yes. It is when it translate the JSON to ChooseParameter where the method signature have Object and no converter has been registered for Object

Comment (jglick) [reply src/test/java/jenkins/branch/ParameterDefinitionBranchPropertyTest.java:141]: Yes, just curious whether this caused the entire form submission to fail (HTTP 500 or similar), or if a stack trace got dumped on the system log and the bean was just left unconfigured.

Comment (nfalco79) [reply src/test/java/jenkins/branch/ParameterDefinitionBranchPropertyTest.java:141]: for what I remember I got the error page with stacktrack with the jenkins picture in evil mode

File: src/test/java/jenkins/branch/ParameterDefinitionBranchPropertyTest.java [hunk 3]
```diff
@@ -116,6 +120,40 @@ public void jobDecorator_Job_implementing_ParameterizedJob() throws Exception {
         assertThat(new ParameterDefinitionBranchPropertyImpl().jobDecorator((Class) ParamJob.class), notNullValue());
     }
 
+    @Issue("JENKINS-61438")
+    // TODO Enable this test case when jenkins.version includes the patch in stapler/stapler#182 
+    @Ignore("Enable this test when stapler/stapler#182 is merged and the jenkins.version will include the patch")
```
Comment (jglick) [inline src/test/java/jenkins/branch/ParameterDefinitionBranchPropertyTest.java:125]:
```suggestion
    @Ignore("TODO enable in 2.242+")
```

**Pull request merged**